### ISSUE TITLE
JedisSentinelPool doesn't switch to another sentinel  when the sentinel is down.

### ISSUE THREAD

**Issue by fangqiao** [xetorthio/jedis#1460]
### Expected behavior

When one sentinel process is down, JedisSentinelPool should automatically connect to another sentinel process that are monitoring the same Reids mater/slave group. 

Did I do anything wrong with configuring the JedisSentinelPool or Redis/Jedis configuration? Or this is the right behavior?

### Actual behavior

But instead it constantly report following error:

一月 16, 2017 5:26:02 下午 redis.clients.jedis.JedisSentinelPool$MasterListener run
严重( severe): Lost connection to Sentinel at 192.168.0.214:26379. Sleeping 5000ms and retrying.
redis.clients.jedis.exceptions.JedisConnectionException: java.net.ConnectException: 拒绝连接 (Connection refused)
	at redis.clients.jedis.Connection.connect(Connection.java:207)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:93)
	at redis.clients.jedis.Connection.setTimeoutInfinite(Connection.java:94)
	at redis.clients.jedis.Jedis.subscribe(Jedis.java:2678)
	at redis.clients.jedis.JedisSentinelPool$MasterListener.run(JedisSentinelPool.java:291)
Caused by: java.net.ConnectException: 拒绝连接 (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at redis.clients.jedis.Connection.connect(Connection.java:184)
	... 4 more

### Steps to reproduce:

1. Configured four sentinel processes to monitor a Redis mater/slave group of 3 nodes.
2. Configured JedisSentinelPool to connect to the four sentinel groups:
  a:  connections = ["192.168.0.214:26379", "192.168.0.214:26380", "192.168.0.214:26381", "192.168.0.214:26382"]
  b: Pool initialization:     val dataSentinelPool: JedisSentinelPool = new JedisSentinelPool(BjConfig.DataCentinel.name,
      BjConfig.DataCentinel.connections.toSet[String].asJava, dataRedisConfig)
  by the way, I use scala. 
3. Kill the primary sentinel
4. Then the application reports the above exception.

### Redis / Jedis Configuration
As stated above.

#### Jedis version:
Latest stable
#### Redis version:
Latest stable 
#### Java version:
1.8

**Comment by sheinbergon** [comment]
@fangqiao That's not how Redis Sentinel Works. 
Sentinel is Database feature for HA in Redis. Various language SDKs ( Java or other ) simplly support it or better said , are "aware" of it.

Determining if the current master is down is up to the sentinel "cluster" ( you can have many sentinel 'servers', you used 4 in your example ). this takes place whenever failover timeout ( as set in the sentinel configuration ) is reached. 

Whenever that happens, the Jedis driver gets notified and reconnects to the new master upon subsequent requests. 

So to sum things up errors are perfectly valid as long as they go away right after the failover tookplace 

If you'd like further assistance , please provide your sentinel configuration :)

**Comment by fangqiao** [comment]
Thanks. It is also what I thought afterwards.

**Comment by marcosnils** [comment]
@sheinbergon thx for taking care!

**Comment by etlweather** [comment]
This is old but I don't really get this answer. If a Redis master goes away, the Redis client ask Sentinel for the new master - that make sense.

But if the client is connected to Sentiel A and it goes away, the client should connect to Sentinel B. 

This is covered in the [Guidelines for Redis clients with support for Redis Sentinel](https://redis.io/topics/sentinel-clients)

> ### Step 1: connecting to the first Sentinel
>   
> The client should iterate the list of Sentinel addresses. For every address it should try to connect to the Sentinel, using a short timeout (in the order of a few hundreds of milliseconds). On errors or timeouts the next Sentinel address should be tried.
> 
> If all the Sentinel addresses were tried without success, an error should be returned to the client.
> 
> The first Sentinel replying to the client request should be put at the start of the list, so that at the next reconnection, we'll try first the Sentinel that was reachable in the previous connection attempt, minimizing latency.

I am interested in this because I just ran into this scenario. The client was still connected to the Redis master and everything was working, but it kept logging that connection error.

`Lost connection to Sentinel at xyz.example.com:26379. Sleeping 5000ms and retrying.`

That does not seem like the correct behavior, particularly that there are 5 Sentinel servers provided in the configuration (at least, provided via Spring Boot for Spring Session - I didn't check Spring code to see what they do with that).

**Comment by etlweather** [comment]
OK, it seems like this log entry is somewhat of a red-herring - it leads one to think something is wrong, but it may not be - if you have multiple Sentinels.

From looking at the code, what I see is you start a connection to every Sentinels listed. So if the connection to a sentinel is not working, you still have connections to the other sentinels and should receive a master change event, if the Redis-server master was to change.

I would have to crash the redis-server master while I crash a sentinel as well to full test my conclusion from the code.

**Comment by soloestoy** [comment]
> If a Redis master goes away, the Redis client ask Sentinel for the new master - that make sense.

Just wonder if jedis does support this feature now? @etlweather @marcosnils @sheinbergon
